### PR TITLE
[WIP] Temporal improvements

### LINKED ISSUE #219: GeoTrellisRasterSource temporal build improvements
At this point [GeoTrellisRasterSource.build function](https://github.com/geotrellis/geotrellis-server/blob/571ea30483b72f643acbbea031e66071cd9680e9/core/src/main/scala/geotrellis/store/GeoTrellisRasterSourceLegacy.scala#L270-L299) creates a set of GeoTrellisRasterSources from a temporal layer, making an assumption that for each time bin there is a single spatial slice. This is not true in the most cases in real world:

1. Instead of an optional temporal resolution we need to pass a list of dates that we know are present in the layer,
2. Have an optional ability to read the list of dates that correspond to each temporal slice from the AttributeStore (this can be done as some custom attribute saved into the AttributeStore).
3. Creation of each new GeoTrellisRasterSource for each date is expensive now, since for each new instance we read all the metadata everytime for each slice, though it is the same across all of them.

### PR DESCRIPTION
**Pull request by CloudNiner**
## Overview

**My comments are in the git log for the HEAD commit on this branch. Additional context there.**

This PR addresses the major concerns raised in #219:
- [x] Can we revert to upstream GeoTrellisRasterSource? Yes -- we'll need to add the time related changes to upstream though
- [x] Can we minimize the number of RasterSource instantiations we do for GeoTrellis layers? Yes -- Now 2. One when we build the conf on server start, one every time we go to read from the layer in a GetMap or equivalent call
- [x] Can we defer RasterSource instantiation until just before we're ready to read data? Yes, see bullet above

### Checklist

- [x] Description of PR is in an appropriate section of the CHANGELOG and grouped with similar changes if possible

## Testing Instructions

It should still be possible to load and view both application.conf. It should still be possible to load application-spacetimekey.conf and view the GetCapabilities or equivalent views for WMS/WCS manually. GetMap or equivalent views will still time out due to excessive reads of the underlying test rastersource.

Closes #219

### CODE REVIEW

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    timeMetadataKey: String = "times"
+) extends RasterOgcSource {
+
+  private val dataPath = GeoTrellisPath.parse(sourceUri)
+
+  // If layer is temporally enabled, default time is the first in the layer
+  lazy val source =
+    GeoTrellisRasterSourceLegacy(dataPath, timeInterval.map(_.start))
+
+  lazy val times: Option[NEL[ZonedDateTime]] = {
```
Comment (echeipesh) [inline ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: Optional NonEmptyList, so is that like a list?
Didn't trace the semantics of this this but the signature made me laugh.

Comment (pomadchin) [reply ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: `List` 💯

File: ogc/src/main/scala/geotrellis/server/ogc/wms/WmsModel.scala [hunk 2]
```diff
@@ -16,62 +16,78 @@
...
+            p.styles.headOption
+              .filterNot(_.isEmpty)
+              .orElse(source.defaultStyle)
+          val style: Option[OgcStyle] =
+            styleName.flatMap { name =>
+              source.styles.find(_.name == name)
+            }
+          source match {
+            case MapAlgebraSource(name,
+                                    title,
```
Comment (pomadchin) [inline ogc/src/main/scala/geotrellis/server/ogc/wms/WmsModel.scala]: How do you feel about such formating? Looks a bit unreadble to me ._.

Also it looks like the old `for` syntax was okay; if you don't like it how about smth like:
image

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            case OgcTimeInterval(start, Some(end), _) => start <= t && t < end
+            case _                                    => false
+          }
+        }
+        .fold(source)(sourceForTime(_))
+    }
+  }
+
+  def sourceForTime(time: ZonedDateTime): GeoTrellisRasterSourceLegacy =
+    GeoTrellisRasterSourceLegacy(dataPath, Some(time))
```
Comment (pomadchin) [inline ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: It looks like there still a new raster sources for a time point is created. Have you decided to move it into a separate PR?

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  def sourceForTime(timeInterval: OgcTimeInterval): GeoTrellisRasterSourceLegacy = {
+    // I'd love some suggestions for how to write this better...
+    times.fold(source) { timeList =>
+      timeList
+        .find { t =>
+          timeInterval match {
+            case OgcTimeInterval(start, None, _)      => start == t
+            case OgcTimeInterval(start, Some(end), _) => start <= t && t < end
```
Comment (pomadchin) [inline ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: Find will return on the first occurrence, is it what we expect?

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }
+
+  lazy val timeInterval: Option[OgcTimeInterval] = times.map {
+    case NEL(head, Nil)  => OgcTimeInterval(head)
+    case NEL(head, tail) => OgcTimeInterval(head, Some(tail.max), None)
+  }
+
+  def sourceForTime(timeInterval: OgcTimeInterval): GeoTrellisRasterSourceLegacy = {
+    // I'd love some suggestions for how to write this better...
```
Comment (pomadchin) [inline ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: It would be smth like 

```
def sourceForTime(timeInterval: OgcTimeInterval): GeoTrellisRasterSourceLegacy = 
    if(times.isEmpty) source
    else times.find { t =>
      timeInterval match {
        case OgcTimeInterval(start, None, _) => start == t
        case OgcTimeInterval(start, Some(end), _) => start <= t && t < end
        case _ => false
      }
    }.fold(source)(sourceForTime)
```

After you'll change the `times` signature.

**Comment by CloudNiner** [comment]
>  Is it expected that in this PR there was nothing done to optimize GeoTrellisRasterSources creation

So the change to address that is here: https://github.com/geotrellis/geotrellis-server/pull/234/files#diff-fb4892c4b98bbccefb508b128e997082R38

Previously when the server started we were creating a new RasterSource for all possible times in the GtRasterSource (via the now removed `build()` function) + the parent raster source. Now we create the parent RasterSource on server start to get the valid times out of it, and then only create a single RasterSource valid for the request time when a user makes a temporal request. Requests with no temporal component just use the parent RasterSource defined at `GeoTrellisOgcSource.source`. This seems like the best we can do...at some point we have to create a raster source valid at the appropriate time in order to request the data. Is there some further optimization we can make?

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala (hunk 5 above)
Comment (CloudNiner) [reply ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: LGTM, updated

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala (hunk 1 above)
Comment (CloudNiner) [reply ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: Switched to `List`

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala (hunk 4 above)
Comment (CloudNiner) [reply ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: Yeah. Since we can only render a raster valid for a single time, we just pick the first one in the list that matches the interval. I added a `.sorted` to `lazy val times` so that we always pick the time closest to the start interval if there are multiple matches. That way there's some consistency.

File: ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala (hunk 3 above)
Comment (CloudNiner) [reply ogc/src/main/scala/geotrellis/server/ogc/source/GeoTrellisOgcSource.scala]: I _think_ this is the same comment [as below](https://github.com/geotrellis/geotrellis-server/pull/234#issuecomment-603477419). Correct me if I'm wrong?

File: ogc/src/main/scala/geotrellis/server/ogc/wms/WmsModel.scala (hunk 2 above)
Comment (CloudNiner) [reply ogc/src/main/scala/geotrellis/server/ogc/wms/WmsModel.scala]: Oof. Yeah the formatting here sucks. I hate that the intellij scalafmt formatter defaults to indenting parameters like that.

Updated

File: ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala [hunk 6]
```diff
@@ -75,14 +69,73 @@ case class SimpleSource(
...
+  lazy val source =
+    GeoTrellisRasterSourceLegacy(dataPath, timeInterval.map(_.start))
+
+  lazy val isTemporal: Boolean = !times.isEmpty
+
+  lazy val times: List[ZonedDateTime] = {
+    val layerId = source.dataPath.layerId
+    val header = source.attributeStore.readHeader[LayerHeader](layerId)
+    if (header.keyClass.contains("SpaceTimeKey")) {
+      source.attributeStore.read[List[ZonedDateTime]](layerId, timeMetadataKey).sorted
```
Comment (pomadchin) [inline ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala]: Add an import `geotrellis.store.query.CirceCompat._` in all places you use temporal circe codecs. The sotry behind that is that Scala 2.12 circe handles temporal codecs in a differnet way (they are supproted by default) and in early circe versions such codecs lived in a separate circe module.

Comment (pomadchin) [reply ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala]: @CloudNiner ^ this is a comment that explains why Scala 2.11 builds are failing.

File: ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala [hunk 7]
```diff
@@ -77,14 +70,75 @@ case class SimpleSource(
...
+  timeMetadataKey: String = "times"
+) extends RasterOgcSource {
+
+  private val dataPath = GeoTrellisPath.parse(sourceUri)
+
+  // If layer is temporally enabled, default time is the first in the layer
+  lazy val source =
+    GeoTrellisRasterSourceLegacy(dataPath, timeInterval.map(_.start))
+
+  lazy val isTemporal: Boolean = !times.isEmpty
```
Comment (pomadchin) [inline ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala]: Actually there is a `.nonEmpty` function :D

File: ogc-example/src/main/resources/application-spacetimekey.conf [hunk 8]
```diff
@@ -190,13 +190,7 @@ layers = {
...
     name = "spacetimetest"
     title = "Landsat 8 OGC Temporal SpaceTimeKey Test"
-    source = {
-      type = "geotrellis"
-      catalog-uri = "s3://geotrellis-test-non-public/spacetimekey-test"
-      layer = "spacetimekey-test"
-      zoom = 14
-      band-count = 1
-    }
+    source = "gt+s3://geotrellis-test-non-public/spacetimekey-test?layer=spacetimekey-test&zoom=14&band_count=1"
```
Comment (pomadchin) [inline ogc-example/src/main/resources/application-spacetimekey.conf]: Also it looks like that `ogc-example/src/main/resources/application.conf` has to be changed and tested before the merge.

Comment (CloudNiner) [reply ogc-example/src/main/resources/application-spacetimekey.conf]: Yup. Addressed.

File: ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala (hunk 6 above)
Comment (pomadchin) [reply ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala]: Ok, [here is the problem](https://github.com/locationtech/geotrellis/blob/0e60b860825d289f7283bea9683f3ca4b2f3c4c5/store/src/main/scala/geotrellis/store/json/Implicits.scala#L95-L100) we need to remove it from the GeoTrellis core.

**Comment by pomadchin** [comment]
Create a new GeoTrellis issue https://github.com/locationtech/geotrellis/issues/3208

File: ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala (hunk 6 above)
Comment (CloudNiner) [reply ogc/src/main/scala/geotrellis/server/ogc/OgcSource.scala]: Issue here resolved by removing the CirceCompat import from GeoTrellisRasterSourceLegacy which pulls a ZonedDateTime decoder from the package listed in the link above.

**Comment by CloudNiner** [comment]
Whoops, missed the `rastersourceconf` change. I updated CHANGELOG to note that before I merged.

**Comment by pomadchin** [comment]
Docs live here https://github.com/geotrellis/geotrellis-server/blob/develop/ogc-example/docs/conf.md#temporal-layers (adding this link for a better discoverability)
